src/include/sof/audio/buffer.h

```diff
@@ -145,6 +145,31 @@ void comp_update_buffer_produce(struct comp_buffer *buffer, uint32_t bytes);
 /* called by a component after consuming data from this buffer */
 void comp_update_buffer_consume(struct comp_buffer *buffer, uint32_t bytes);
 
+static inline void buffer_lock(struct comp_buffer *buffer)
+{
+	if (!buffer->inter_core)
+		return;
+
+	spin_lock(buffer->lock);
+
+	/* invalidate in case something has changed during our wait */
+	dcache_invalidate_region(buffer, sizeof(*buffer));
+}
+
+static inline void buffer_unlock(struct comp_buffer *buffer)
+{
+	if (!buffer->inter_core)
+		return;
+
+	/* save lock pointer to avoid memory access after cache flushing */
+	spinlock_t *lock = buffer->lock;
+
+	/* wtb and inv to avoid buffer locking in read only situations */
+	dcache_writeback_invalidate_region(buffer, sizeof(*buffer));
```

Comment:
do you really need to invalidate the caches both in _lock and in _unlock?

Reply:
Can we also comment on what these function do/permit t the high level (at the start if each).

Reply:
@tlauda could you please comment on this? Additionally, maybe this question was already asked and answered in one of discussions, marked as "resolved" and therefore closed above, sorry if that's the case. But I'm still unclear how about this this works: you start a pipeline processing on core 0, it processes components one by one, until it reaches component A, at which stage it requests core 1 to process that component. But what does core 0 do during that time? Just waits for core 1 to complete? Or processing asynchronous events like IPC? But usually there aren't many of those during streaming? What are we gaining?

Reply:
@lyakh I think I have already did in different comment. It invalidates to avoid calling buffer_lock/unlock in params and prepare calls, where nothing inside buffer is modified, but only read.
Core 0 only processes it's own components and doesn't wait for core 1. Processing on both cores happens totally asynchronous.

Reply:
@tlauda you're probably right, but sorry I still don't understand. I thought data processing in audio pipelines was linear: the next consumer can only start after the previous producer is done. I could understand if the other-core component was the first or the last in the pipeline, e.g.
```
core 0      core 1
A0
B0
A1            C0
B1
A2            C1
```
i.e. the original core can start processing the next data block while the "offload core" finishes processing the previous one. Is this what is used in this PR? I see that @lgirdwood is satisfied with this PR, so, it's probably just me who doesn't understand.

Reply:
@lyakh It's not always linear. We're going to skip components, which are not ours (owner is different core). This way other core can in the same time process the component inside our pipeline and with the next pipeline_copy we will have data to work on.

Reply:
@lyakh Maybe there is some data from a _previous_ run of the pipeline that the next component can use. Sure, we'd have a bit of extra latency (1ms per component that is on a different core) but that may be acceptable.

So Host -> ParallelProcessing -> Dai:

First time Host would get the data and ParallelProcessing would be scheduled. DAI would have nothing to do.
Next millisecond Host would get new data, ParallelProcessing will have finished the previous ms and will be scheduled for the new data, DAI would get the old data from the previous ms.